core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -681,4 +642,120 @@ private void doRelease() {
+                return;
+            }
+
+            // Invoke the notification task immediately if complete already.
+            notifyChild(desiredProtocol, key, childPromise, timingsBuilder,
+                        isCompletedExceptionally() ? null : getNow(null));
+        }
+
+        private void notifyChild(SessionProtocol desiredProtocol, PoolKey key,
+                                 ChannelAcquisitionFuture childPromise,
+                                 ClientConnectionTimingsBuilder timingsBuilder,
+                                 @Nullable PooledChannel pch) {
+
+            // 1 - use the current pending acquisition.
+            // 2 - create a new connection.
+            // 3 - use another pending acquisition.
+            final int result;
+            if (pch != null) {
+                final SessionProtocol actualProtocol = pch.protocol();
+                if (actualProtocol.isMultiplex()) {
+                    final HttpSession session = HttpSession.get(pch.get());
+                    if (session.incrementNumUnfinishedResponses()) {
+                        result = 1;
+                    } else if (usePendingAcquisition(actualProtocol, key, childPromise, timingsBuilder)) {
+                        result = 3;
+                    } else {
+                        result = 2;
+                    }
+                } else {
+                    // Try to acquire again because the connection was not HTTP/2.
+                    // We use the exact protocol (H1 or H1C) instead of 'desiredProtocol' so that
+                    // we do not waste our time looking for pending acquisitions for the host
+                    // that does not support HTTP/2.
+                    final PooledChannel ch = acquireNow(actualProtocol, key);
+                    if (ch != null) {
+                        pch = ch;
+                        result = 1;
+                    } else {
+                        result = 2;
+                    }
+                }
+            } else {
+                result = 2;
+            }
+
+            switch (result) {
+                case 1:
+                    timingsBuilder.pendingAcquisitionEnd();
+                    childPromise.complete(pch);
+                    break;
+                case 2:
+                    timingsBuilder.pendingAcquisitionEnd();
+                    connect(desiredProtocol, key, childPromise, timingsBuilder);
+                    break;
+                case 3:
+                    // There's nothing to do because usePendingAcquisition() was called successfully above.
+                    break;
+            }
+        }
+
+        @Override
+        public boolean complete(PooledChannel value) {
+            assert value != null;
+            if (!super.complete(value)) {
+                return false;
+            }
+
+            final List<Consumer<PooledChannel>> notifications = this.notifications;
```

Comment:
Think we can extract a helper for this and `completeExceptionally` code fragments